README.md
# Jspxcms

本源码包采用maven结构，和普通eclipse的web项目结构不一样，且不包含lib目录下的第三方jar包。

## 环境要求

- JDK7或更高版本（支持JDK8）。建议使用JDK8，有更好的内存管理。更低版本的JDK7可能需要设置Java内存`-XX:PermSize=128M -XX:MaxPermSize=512M`，否则可能出现这种类型的内存溢出：`java.lang.OutOfMemoryError: PermGen space`。
- Servlet3.0或更高版本（如Tomcat7或更高版本）。
- MySQL5.5或更高版本（如需使用MySQL5.0，可将mysql驱动版本替换为5.1.24）；Oracle10g或更高版本；SQLServer2005或更高版本。
- Maven3.2或更高版本。
- 系统后台兼容的浏览器：IE9+、Edge、Firefox、Chrome。
- 前台页面兼容的浏览器取决于模板，使用者可以完全控制模板，理论上可以支持任何浏览器。自带的默认模板兼容的浏览器：IE8+、Edge、Firefox、Chrome。

## 搭建步骤

1. 创建数据库。如使用MySQL，字符集选择为`utf8`或者`utf8mb4`（支持更多特殊字符如表情字符emoji，推荐）。
2. 执行数据库脚本。数据库脚本在`database`目录下。
3. 在eclipse中导入maven项目。点击eclipse菜单`File` - `Import`，选择`Maven` - `Existing Maven Projects`。创建好maven项目后，会开始从maven服务器下载第三方jar包（如spring等），需要一定时间，请耐心等待。
4. 修改数据库连接。打开`/src/main/resources/application.propertis`文件，根据实际情况修改`spring.datasource.url`、`spring.datasource.username`、`spring.datasource.password`的值。
5. 启动程序。**请一定按照这里介绍的方式启动程序，否则可能因上下文路径（ContextPath）导致页面没有样式等问题**。在eclipse中，右键点击项目名，选择`Run as` - `Maven build...`，`Goals`填入`spring-boot:run`，`JRE`选项卡中可以选择JDK版本，`VM arguments`输入框中可以设置java内存参数，如使用JDK7，应在此处填入`-XX:PermSize=128M -XX:MaxPermSize=512M`，然后点击`Run`。
6. 访问系统。前台地址：[http://localhost:8080/](http://localhost:8080/)，使用手机访问前台或者使用浏览器模拟手机访问前台，系统会调用手机端模板，显示手机端的界面。后台地址：[http://localhost:8080/cmscp/index.do](http://localhost:8080/cmscp/index.do)，用户名：admin，密码：空。

## Maven镜像

使用国外的Maven仓库速度很慢、甚至连接不上，很容易出错。可以使用aliyun的Maven镜像仓库。

修改maven的settings.xml文件，mirrors部分使用以下配置：

	<mirrors>
	  <mirror>
	    <id>central</id>
	    <mirrorOf>*</mirrorOf>
	    <name>Nexus Aliyun</name>
	    <url>http://maven.aliyun.com/nexus/content/groups/public/</url>
	  </mirror>
	</mirrors>

## 上下文路径

程序通常在部署在Tomcat根目录，首页访问地址类似`http://www.mysite.com/` `http://localhost/` `http://localhost:8080/`。在一些特殊的场合，如在Eclipse默认的Tomcat启动方式（非maven方式），可能将程序部署在某一个路径下，首页访问地址类似`http://www.mysite.com/jspxcms/` `http://localhost/mysite/` `http://localhost:8080/jspxcms/`。此时访问网站前台会出现样式不能正常显示的情况，可以到后台`系统 - 系统设置`中设置`上下文路径`，类似为`/jspxcms`，其中斜杠`/`不能省略，`jspxcms`为部署目录的路径，如在开发环境，则通常为项目名。

开发环境要避免使用上下文路径，除非网站正式部署时也要部署到相应的路径下，否则在开发环境下上传的图片部署到正式环境时，不能正常显示。因为上传图片时，图片地址会带有上下文路径的信息。

Eclipse默认的tomcat启动方式（非maven方式）会将程序部署到特定目录再启动，并不是直接在项目所在目录启动tomcat，这时上传的图片（包括通过系统后台新增和修改的模板）也保存在特定的部署目录，并不会保存在程序所在的目录。当修改了Eclipse源代码或文件，会自动重新部署程序，之前上传的图片会被清空。如果发现在开发环境下上传的图片突然都找不到了，很可能就是这个原因。

综上所述，强烈建议使用搭建步骤中介绍的方式启动程序。

## 后端技术

SpringBoot：提供了对Spring开箱即用的功能。简化了Spring配置，提供自动配置 auto-configuration功能。

Spring：是提供了IoC等功能，是目前最流行的Java企业级开发框架。

SpringMVC：MVC框架，使用方便，Bug较少。

JPA：持久化框架。属于JSR标准，JPA实现选择最常用的Hibernate。

SpringDataJPA：对JPA封装，大部分查询只需要在接口中写方法，而不需要实现改方法，极大开发效率。

QueryDSL：实现类型安全的JPA查询，使用对象及属性实现查询，避免编写jpql出现的拼错字符及属性名记忆负担。

FreeMarker：模板组件。

Shiro：安全组件。配置简便。

Lucene：全文检索组件。实现对中文的分词搜索。

Ehcache：缓存组件。主要用在JPA二级缓存、Shiro权限缓存。

Quartz：定时任务组件。

## 前端技术

jQuery：JavaScript库。

Bootstrap：响应式设计前端框架。

AdminLTE：后台管理平台开源框架。

jQuery UI：基于jQuery的UI框架。

jQuery Validation：基于jQuery的表单校验框架。

UEditor：Web富文本编辑器。

Editor.md：基于Markdown语法的Web文本编辑器。

ECharts：用于生成图标的组件。

My97DatePicker：日期组件。

zTree：树组件。



CHANGELOG.txt
# Jspxcms 发版说明

## jspxcms-8.5.0 (2017-10-27)

### 新功能(New Feature)

1. 新增站点间推送功能。
2. 新增微信jssdk分享功能。
3. 新增站点导入导出功能。

### 改进(Improvement)

1. 站点复制功能可以复制所有模板、模型、栏目和文档。
2. 站点独立配置微信公众号。
3. 将flash上传组件改为jquery上传组件。

## jspxcms-8.0.3 (2017-10-25)

### 改进(Improvement)

1. 安全升级。

### 缺陷(Bug)

2. 修复：后台首页中上次登录时间不更新。


## jspxcms-8.0.2 (2017-09-29)

### 改进(Improvement)

1. 后台登录页面禁止密码自动完成。否则浏览器会提示是否保存密码，部分安全扫描软件会认定为漏洞。

### 缺陷(Bug)

1. 修复：系统通知和我的私信在oracle下报错。
2. 修复：工作流步骤修改时报错。
3. 修复：删除有嵌套评论（评论盖楼）的文章报错。
4. 修复：栏目移动报错。
5. 修复：栏目合并提交报错。
6. 修复：前台信息提示页nextUrl的跨站攻击问题。
7. 修复：后台登录页面用户名跨站攻击问题。


## jspxcms-8.0.1 (2017-06-21)

### 改进(Improvement)

1. 在站点设置里可以增加手机模板修改，不用到站点管理里修改。
2. sql脚本增加commit语句，避免因为执行脚本时为提交导致数据没有正确插入。
3. pom.xml里增加tomcat7启动插件，方便希望使用tomcat7插件启动的用户。
4. 文档管理自动获取描述长度增加到最长的450字符。
5. 友情链接列表logo限制大小，避免logo尺寸过大导致页面显示不好看。
6. InfoPage和NodePage标签增加p0参数。
7. 升级MySQL驱动，支持MySQL5.7。

### 缺陷(Bug)

1. 修复：后台右上角小房子图标的前台首页链接在多站点情况下，切换站点后无变化。
2. 修复：文档管理标题图、属性图、图集在火狐下点击上传按钮无反应。
3. 修复：文件管理搜索功能没有保留当前文件夹路径。
4. 修复：后台左侧菜单没有屏蔽无权限的菜单。
5. 修复：开启页面静态化后，某些栏目生成了HTML，但还会显示HTML待更新。
6. 修复：文档管理doc上传按钮位置不正确。
7. 修复：有上下文路径时，站点静态首页地址多了一个上下文路径。
8. 修复：谷歌某些版本的浏览器点击文档管理图片选择按钮无效。
9. 修复：后台任务管理翻页列表固定不变。
10. 修复：修改组织报错。


## jspxcms-8.0.0 (2017-04-19)

### 新功能(New Feature)

1. 使用springboot框架。
2. 前台使用restful风格地址，去掉url地址后缀。
3. 支持多平台访问。同一站点可分别指定PC端模板和手机端模板。并支持PC端、手机端同时生成静态页。
4. 新增一键启用、关闭静态化功能。
5. 后台界面美化。
6. 新增用户模型。
7. 前台增加站内信、通知、收藏功能。
8. 升级到JDK7、Servlet3.0（Tomcat7或更高版本）。

### 改进(Improvement)

1. 前台会员投稿可发布图片集、视频等内容。
2. 后台菜单使用yaml格式文件存储，结构更清晰，维护更方便。
3. 去掉URL带域名设置，勾选域名识别则自动URL带域名。
4. 去掉安装功能。部署时手动导入数据。

### 升级指南（Upgrade Guide）

1. 搜索模板的`.jspx'` `.jspx"` `.jspx%` `.jspx;` `.jspx&` `.jspx<` `.jspx?`（如直接搜索`.jspx`，会匹配到`.jspxcms...`），替换掉其中的`.jspx`。
2. 搜索模板的`.servlet'` `.servlet"` `.servlet?`，替换掉其中的`.servlet`。
3. 取消手动设置默认站点，站点ID为1的为默认站点。
4. 系统后台浏览器支持IE9+，需要使用IE8的用户请使用7.X版本。


## jspxcms-7.0.2-release (2017-03-24)

### 改进(Improvement)

1. 文章正文中小于100px的图片不作为标题图。
2. 会员中心头像图片增加时间戳参数，避免因为缓存导致修改头像后无变化。
3. 创建站点时，同时复制首页栏目，以免在未创建首页栏目时，访问首页出现node not found: root。
4. 删除站点时，同时删除空栏目。

### 缺陷(Bug)

1. 修复：调查问卷限制会员投票无效。
2. 修复：设置了编辑器图片最大宽度，而最大高度为0是会报错。
3. 修复：会员中心上传头像时，不裁剪图片直接提交会报错。
4. 修复：部分关联表数据删除后，index顺序未调整，有可能导致文档或用户相关功能报错。
5. 修复：删除站点时，提示违反数据约束。


## jspxcms-7.0.1-release (2017-01-20)

### 改进(Improvement)

1. 访问统计中，饼图改进为只显示前10位，剩余的作为其他显示。

### 缺陷(Bug)

1. 修复：文档管理修改时分部默认数据会报错的问题（数据库初始化数据导致的问题）。
2. 修复：访问统计中的菜单没有全部加上。
3. 修复：访问统计中，URL地址包含空格会报错。
4. 修复：获取部分IP的实际地址会报错。
5. 修复：搜索中文时，出现乱码。


## jspxcms-7.0.0-release (2016-12-26)

### 新功能(New Feature)

1. 文档录入增加doc导入功能，可以导入doc里面的图片，避免一张张上传。
2. 增加找回密码邮件的有效期，超过8小时失效。
3. 增加角色等级，用户不可以被赋予比自身等级更高的角色，避免拥有用户管理权限的人员提升自己的权限。
4. 采集增加是否采集正文图片选项。
5. 增加栏目权限控制，用户无法访问没有权限的栏目。
6. 增加通过IP地址获取真实位置的功能。
7. 后台浏览统计通过图表显示。
8. 后台增加站内信功能。
9. 支持使用ImageMagick或GraphicsMagick处理图片。比纯Java效果更好、效率更高，并支持gif图压缩。

### 改进(Improvement)

1. 使用maven生成QueryDSL查询类。
2. 升级ueditor-1.4.3.3。

### 缺陷(Bug)

1. 修复editor.md编辑器光标定位不准问题。

### 升级指南（Upgrade Guide）

1. **升级ueditor-1.4.3.3**，前台投稿模板中ueditor编辑器调用方式需修改，请参照默认模板同名文件修改`/template/1/default/sys_member_contribution_form.html`。


## jspxcms-6.5.4-release (2016-12-22)

### 缺陷(Bug)

1. 修复：附件管理点击“是否使用”列头排序会报错。
2. 修复：文档管理对HTML过滤会导致一些正常的HTML无法录入。取消过滤。
3. 修复：editormd无法上传图片。
4. 修复：ueditor上传图片最大为4M，以及超过2M的图片会压缩。全部修改为100M。
5. 修复：InfoList(InfoPage)的isPerm参数无效。
6. 修复：前台投稿编辑器上传图片提示网络错误。


## jspxcms-6.5.3-release (2016-11-02)

### 改进(Improvement)

1. 用户不能删除自己。
2. 增加专题的浏览次数功能。
3. 图集原图如不需要压缩则不要处理，避免导致gif图动态效果消失。

### 缺陷(Bug)

1. 修复：微信登录没有显示微信名，微博登录StateIllegalException异常。
2. 修复：投稿修改的正文直接显示html。
3. 修复：发送邮件报错，没有mail的实现jar包。
4. 修复：采集是否允许标题重复，没有验证、没有默认值。
5. 修复：没有用户批量删除权限的未屏蔽按钮。
6. 修复：前台头像剪裁在特定情况下由于浮点运算误差，导致剪裁高度大于原图，出现异常。
7. 修复：文库转换不支持docx。


## jspxcms-6.5.2-release (2016-08-10)

### 改进(Improvement)

1. MySQL驱动改为5.1.24，兼容MySQL5.1。
2. 游客组、匿名用户、根用户不能被删除。
3. 默认模板cover_news.html不固定获取news栏目的子栏目。
  
### 缺陷(Bug)

1. 修复：调查问卷-修改调查项-保存并返回，出现404错误。


## jspxcms-6.5.1-release (2016-07-25)

### 改进(Improvement)

1. 敏感词的替换词允许为空。
2. javassist版本降低为hibernate依赖的版本3.18.1，避免部分小版本较低的tomcat报错。
  
### 缺陷(Bug)

1. 修复：文档管理产品介绍正文没有显示内容。
2. 修复：新建站点后，创建首页栏目报错。
3. 修复：采集管理，新增报错。


## jspxcms-6.5.0-release (2016-07-07)

### 新功能(New Feature)

1. 开源版开放多站点功能。
2. 增加手机端网站演示。
3. 增加微信公众号群发新闻功能。
4. 增加微信第三方登录。
4. 文档管理增加markdown编辑器。
5. 文档管理加上按发布人、作者搜索。
6. 采集时，允许勾选采集标题重复的内容。

### 改进(Improvement)

1. 使用maven管理。
2. 第三方类库版本升级。servlet升级为2.5，spring升级为4.2.x，jpa升级为2.1，hibernate升级为5.0.x，spring-data-jpa升级为1.9.x。
3. 删除文档相关数据时，如果回收站有数据，则自动删除。
4. 删除站点时，自动删除模型。
5. 后台发生400错误时，提供更详细的信息，更便于二次开发时查找问题。
6. 裁剪不删除原图片。
	
### 缺陷(Bug)

1. 前台swfupload增加jsessionid，防止上传时无法获得用户身份信息。
2. “系统设置 - 注册设置”里会员组选项不显示“游客组”。
3. 修复前台登录时，如果上下文路径不为空，则返回地址会出错。
4. 默认模版/template/1/default/info_video.html，42行，`c:0，`改为`c:0,`（中文逗号改为英文逗号）。
5. 修复默认会员中心模版修改头像页面的标题显示为修改邮箱。
6. 修复删除站点时，任务数据未删除。
7. 修复广告图片上传会被标识为未使用。

### 升级指南（Upgrade Guide）

1. **第三方登录URL地址修改。**`/oauth/qq_login.jspx`改为`/oauth/login/qq.jspx`；`/oauth/weibo_login.jspx`改为`/oauth/login/weibo.jspx`。
2. **编辑器工具条`tarbar_Cms`改为`tarbar_Standard`。**
3. **专题类别模版和URL地址修改。**`/specials.jspx`改为`/special_category.jspx`，`/specials/{id}.jspx`改为`/special_category/{id}.jspx`，`sys_specials.html`改为`sys_special_category_list.html`，`sys_specials_list.html`改为`sys_special_category.html`。
4. **Tag模版和URL地址修改。**`/tags.jspx`改为`/tag.jspx`，`/tags/{id}.jspx`改为`/tag/{id}.jspx`；`sys_tags.html`改为`sys_tag_list.html`。
5. **配置文件路径修改。**`/WEB-INF/conf`改为`/WEB-INF/classes/conf`，**国际化文件路径**`/WEB-INF/messages`改为`/WEB-INF/classes/messages`，**数据库连接文件**`/WEB-INF/custom.propertis`改为`/WEB-INF/classes/custom.propertis`。
6. **执行数据库升级脚本：upgrade-6.0to6.5.sql。**


## jspxcms-6.0.1-release (2016-02-22)

### 改进(Improvement)

1. 第三方登录完善账户信息时允许不填写密码和验证码。
2. 广告管理图片为必填项。
3. 修正新浪采集演示数据。
	
### 缺陷(Bug)

1. 文件管理列表页面底部显示图片时，图片定位不准。
2. 广告管理新增广告时报错。
3. oracle版文档管理新增文档，如保存正文为空，更新正文不为空，则cms_info_clob表会出现两条记录。
4. 栏目移动提交时出现404错误。


## jspxcms-6.0.0-release (2015-12-25)

### 新功能(New Feature)

1. 调查问卷。
2. 会员头像、会员空间。
3. 评论盖楼、评论点赞。
4. QQ、微博第三方登录。
5. 高级搜索，关键词高亮。
6. 可查询字段演示。
7. 文档、栏目可查询字段增加p0。
8. 缓存的刷新触发设置。
9. 登录需要验证码的错误次数设置。
10. 演示模版升级。
	

src/main/webapp/park/js/common/toLunar.js
/**用法
 * Lunar.toSolar(2016, 6, 3); 农历转化公历
 * Lunar.toLunar(2016, 7, 6); 公历转化农历
 */
var Lunar = {
  MIN_YEAR : 1891,
  MAX_YEAR : 2100,
  lunarInfo : [
    [0,2,9, 21936], [6,1,30, 9656], [0,2,17, 9584], [0,2,6, 21168], [5,1,26,43344], [0,2,13,59728],
    [0,2,2, 27296], [3,1,22,44368], [0,2,10,43856], [8,1,30,19304], [0,2,19,19168], [0,2,8, 42352],
    [5,1,29,21096], [0,2,16,53856], [0,2,4, 55632], [4,1,25,27304], [0,2,13,22176], [0,2,2, 39632],
    [2,1,22,19176], [0,2,10,19168], [6,1,30,42200], [0,2,18,42192], [0,2,6, 53840], [5,1,26,54568],
    [0,2,14,46400], [0,2,3, 54944], [2,1,23,38608], [0,2,11,38320], [7,2,1, 18872], [0,2,20,18800],
    [0,2,8, 42160], [5,1,28,45656], [0,2,16,27216], [0,2,5, 27968], [4,1,24,44456], [0,2,13,11104],
    [0,2,2, 38256], [2,1,23,18808], [0,2,10,18800], [6,1,30,25776], [0,2,17,54432], [0,2,6, 59984],
    [5,1,26,27976], [0,2,14,23248], [0,2,4, 11104], [3,1,24,37744], [0,2,11,37600], [7,1,31,51560],
    [0,2,19,51536], [0,2,8, 54432], [6,1,27,55888], [0,2,15,46416], [0,2,5, 22176], [4,1,25,43736],
    [0,2,13, 9680], [0,2,2, 37584], [2,1,22,51544], [0,2,10,43344], [7,1,29,46248], [0,2,17,27808],
    [0,2,6, 46416], [5,1,27,21928], [0,2,14,19872], [0,2,3, 42416], [3,1,24,21176], [0,2,12,21168],
    [8,1,31,43344], [0,2,18,59728], [0,2,8, 27296], [6,1,28,44368], [0,2,15,43856], [0,2,5, 19296],
    [4,1,25,42352], [0,2,13,42352], [0,2,2, 21088], [3,1,21,59696], [0,2,9, 55632], [7,1,30,23208],
    [0,2,17,22176], [0,2,6, 38608], [5,1,27,19176], [0,2,15,19152], [0,2,3, 42192], [4,1,23,53864],
    [0,2,11,53840], [8,1,31,54568], [0,2,18,46400], [0,2,7, 46752], [6,1,28,38608], [0,2,16,38320],
    [0,2,5, 18864], [4,1,25,42168], [0,2,13,42160], [10,2,2,45656], [0,2,20,27216], [0,2,9, 27968],
    [6,1,29,44448], [0,2,17,43872], [0,2,6, 38256], [5,1,27,18808], [0,2,15,18800], [0,2,4, 25776],
    [3,1,23,27216], [0,2,10,59984], [8,1,31,27432], [0,2,19,23232], [0,2,7, 43872], [5,1,28,37736],
    [0,2,16,37600], [0,2,5, 51552], [4,1,24,54440], [0,2,12,54432], [0,2,1, 55888], [2,1,22,23208],
    [0,2,9, 22176], [7,1,29,43736], [0,2,18, 9680], [0,2,7, 37584], [5,1,26,51544], [0,2,14,43344],
    [0,2,3, 46240], [4,1,23,46416], [0,2,10,44368], [9,1,31,21928], [0,2,19,19360], [0,2,8, 42416],
    [6,1,28,21176], [0,2,16,21168], [0,2,5, 43312], [4,1,25,29864], [0,2,12,27296], [0,2,1, 44368],
    [2,1,22,19880], [0,2,10,19296], [6,1,29,42352], [0,2,17,42208], [0,2,6, 53856], [5,1,26,59696],
    [0,2,13,54576], [0,2,3, 23200], [3,1,23,27472], [0,2,11,38608], [11,1,31,19176],[0,2,19,19152],
    [0,2,8, 42192], [6,1,28,53848], [0,2,15,53840], [0,2,4, 54560], [5,1,24,55968], [0,2,12,46496],
    [0,2,1, 22224], [2,1,22,19160], [0,2,10,18864], [7,1,30,42168], [0,2,17,42160], [0,2,6, 43600],
    [5,1,26,46376], [0,2,14,27936], [0,2,2, 44448], [3,1,23,21936], [0,2,11,37744], [8,2,1, 18808],
    [0,2,19,18800], [0,2,8, 25776], [6,1,28,27216], [0,2,15,59984], [0,2,4, 27424], [4,1,24,43872],
    [0,2,12,43744], [0,2,2, 37600], [3,1,21,51568], [0,2,9, 51552], [7,1,29,54440], [0,2,17,54432],
    [0,2,5, 55888], [5,1,26,23208], [0,2,14,22176], [0,2,3, 42704], [4,1,23,21224], [0,2,11,21200],
    [8,1,31,43352], [0,2,19,43344], [0,2,7, 46240], [6,1,27,46416], [0,2,15,44368], [0,2,5, 21920],
    [4,1,24,42448], [0,2,12,42416], [0,2,2, 21168], [3,1,22,43320], [0,2,9, 26928], [7,1,29,29336],
    [0,2,17,27296], [0,2,6, 44368], [5,1,26,19880], [0,2,14,19296], [0,2,3, 42352], [4,1,24,21104],
    [0,2,10,53856], [8,1,30,59696], [0,2,18,54560], [0,2,7, 55968], [6,1,27,27472], [0,2,15,22224],
    [0,2,5, 19168], [4,1,25,42216], [0,2,12,42192], [0,2,1, 53584], [2,1,21,55592], [0,2,9, 54560]
  ],
  //是否闰年
  isLeapYear : function(year) {
    return ((year % 4 == 0 && year % 100 != 0) || (year % 400 == 0));
  },
  //天干地支年
  lunarYear : function(year) {
    var gan = ['庚', '辛', '壬', '癸', '甲', '乙', '丙', '丁', '戊', '己'],
      zhi = ['申', '酉', '戌', '亥', '子', '丑', '寅', '卯', '辰', '巳', '午', '未'],
      str = year.toString().split("");
    return gan[str[3]] + zhi[year % 12];
  },
  //生肖年
  zodiacYear : function(year) {
    var zodiac = ['猴', '鸡', '狗', '猪', '鼠', '牛', '虎', '兔', '龙', '蛇', '马', '羊'];
    return zodiac[year % 12];
  },
  //公历月份天数
  //@param year 阳历-年
  //@param month 阳历-月
  solarMonthDays : function(year, month) {
    var FebDays = this.isLeapYear(year) ? 29 : 28;
    var monthHash = ['', 31, FebDays, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    return monthHash[month];
  },
  //农历月份天数
  lunarMonthDays : function(year, month) {
    var monthData = this.lunarMonths(year);
    return monthData[month - 1];
  },
  //农历月份天数数组
  lunarMonths : function(year) {
    var yearData = this.lunarInfo[year - this.MIN_YEAR];
    var leapMonth = yearData[0];
    var bit = (+yearData[3]).toString(2);
    var months = [];
    for (var i = 0; i < bit.length; i++) {
      months[i] = bit.substr(i, 1);
    }
    
    for (var k = 0, len = 16 - months.length; k < len; k++) {
      months.unshift('0');
    }
    
    months = months.slice(0, (leapMonth == 0 ? 12 : 13));
    for (var i = 0; i < months.length; i++) {
      months[i] = +months[i] + 29;
    }
    return months;
  },
  //农历每年的天数
  //@param year 农历年份
  lunarYearDays : function(year) {
    var monthArray = this.lunarYearMonths(year);
    var len = monthArray.length;
    return (monthArray[len-1] == 0 ? monthArray[len-2] : monthArray[len-1]);
  },
  //
  lunarYearMonths : function(year) {
    var monthData = this.lunarMonths(year);
    var res = [];
    var temp = 0;
    var yearData = this.lunarInfo[year - this.MIN_YEAR];
    var len = (yearData[0] == 0 ? 12 : 13);
    for (var i = 0; i < len; i++) {
      temp = 0;
      for (j = 0; j <= i; j++) {
        temp += monthData[j];
      }
      res.push(temp);
    }
    return res;
  },
  //获取闰月
  //@param year 农历年份
  leapMonth : function(year){
    var yearData = this.lunarInfo[year - this.MIN_YEAR];
    return yearData[0];
  },
  //计算农历日期与正月初一相隔的天数
  betweenLunarDays : function(year, month, day) {
    var yearMonth = this.lunarMonths(year);
    var res = 0;
    for (var i = 1; i < month; i++) {
      res += yearMonth[i-1];
    }
    res += day - 1;
    return res;
  },
  //计算2个阳历日期之间的天数
  //@param year 阳历年
  //@param month
  //@param day
  //@param l_month 阴历正月对应的阳历月份
  //@param l_day  阴历初一对应的阳历天
  betweenSolarDays : function(year, month, day, l_month, l_day) {
    var time1 = new Date(year +"-"+ month +"-"+ day).getTime(),
      time2 = new Date(year +"-"+ l_month +"-"+ l_day).getTime();
    return Math.ceil((time1-time2)/24/3600/1000);
  },
  //根据距离正月初一的天数计算阴历日期
  //@param year 阳历年
  //@param between 天数
  lunarByBetween : function(year, between) {
    var lunarArray = [], yearMonth = [], t = 0, e = 0, leapMonth = 0, m = '';
    if (between == 0) {
      t = 1;
      e = 1;
      m = '正月';
    } else {
      year = between > 0 ? year : (year - 1);
      yearMonth = this.lunarYearMonths(year);
      leapMonth = this.leapMonth(year);
      between  = between > 0 ? between : (this.lunarYearDays(year) + between);
      for (var i = 0; i < 13; i++) {
        if (between == yearMonth[i]) {
          t = i + 2;
          e = 1;
          break;
        } else if (between < yearMonth[i]) {
          t = i + 1;
          e = between - ((yearMonth[i-1]) ? yearMonth[i-1] : 0) + 1;
          break;
        }
      }
        
      m = (leapMonth != 0 && t == leapMonth + 1)
      ? ('闰'. this.chineseMonth(t-1))
      : this.chineseMonth(((leapMonth != 0 && leapMonth + 1 < t) ? (t - 1) : t));
    }
    lunarArray.push(year, t, e); //年 月 日
    lunarArray.push(this.lunarYear(year),
            this.zodiacYear(year),
            m,
            this.chineseNumber(e)); //天干地支年 生肖年 月份 日
    lunarArray.push(leapMonth); //闰几月
    return lunarArray;
  },
  //中文月份
  chineseMonth : function(month) {
    var monthHash = ['', '正月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '冬月', '腊月'];
    return monthHash[month];
  },
  //中文日期
  chineseNumber : function(num) {
    var dateHash = ['', '一', '二', '三', '四', '五', '六', '七', '八', '九', '十'];
    if (num <= 10) {
      res = '初'+ dateHash[num];
    } else if (num > 10 && num < 20) {
      res = '十'+ dateHash[num-10];
    } else if (num == 20) {
      res = "二十";
    } else if (num > 20 && num < 30) {
      res = "廿"+ dateHash[num-20];
    } else if (num == 30) {
      res = "三十";
    }
    return res;
  },
  //转换农历
  toLunar : function(year, month, day) {
    var yearData = this.lunarInfo[year - this.MIN_YEAR];
    if (year == this.MIN_YEAR && month <= 2 && day <= 9) {
      return [1891, 1, 1, '辛卯', '兔', '正月', '初一'];
    }
    return this.lunarByBetween(year, this.betweenSolarDays(year, month, day, yearData[1], yearData[2]));
  },
  //转换公历
  //@param year 阴历-年
  //@param month 阴历-月，闰月处理：例如如果当年闰五月，那么第二个五月就传六月，相当于阴历有13个月
  //@param date 阴历-日
  toSolar : function(year, month, day) {
    var yearData = this.lunarInfo[year - this.MIN_YEAR];
    var between = this.betweenLunarDays(year, month, day);
    var ms = new Date(year +"-" + yearData[1] +"-"+ yearData[2]).getTime();
    var s = ms + between * 24 * 60 * 60 * 1000;
    var d = new Date();
    d.setTime(s);
    year = d.getFullYear();
    month = d.getMonth() + 1;
    day  = d.getDate();
    return [year, month, day];
  }
};